### PR TITLE
Configure push notification settings

### PR DESCRIPTION
**Pull request by alber2510**
Do *NOT* merge this PR yet, need testing from MEGAchat side (https://github.com/meganz/MEGAchat/pull/576)

### CODE REVIEW

File: src/megaapi_impl.cpp [hunk 1]
```diff
@@ -6494,6 +6520,75 @@ char *MegaApiImpl::getAvatarColor(handle userhandle)
...
+    int start = mPushSettings->getGlobalScheduleStart();
+    if (start > end) // If start > stop, stop is in the next day, we add 24 hours in minutes
+    {
+        end += 24 * 60;
+    }
+
+    m_time_t timestampNow = m_time(NULL);
+    struct tm structTime;
+    m_localtime(timestampNow, &structTime);
+    int myTimeZone = structTime.tm_gmtoff;
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: The new algorithm doesn't use that value :+1:

File: src/megaapi_impl.cpp [hunk 2]
```diff
@@ -6494,6 +6520,75 @@ char *MegaApiImpl::getAvatarColor(handle userhandle)
...
+        return false;
+    }
+
+    return true;
+}
+
+bool MegaApiImpl::isScheduleNotifiable()
+{
+    if (!mTimezones || !mPushSettings)
+    {
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: Done!

File: src/megaapi_impl.cpp [hunk 3]
```diff
@@ -13863,11 +14053,7 @@ void MegaApiImpl::fetchtimezone_result(error e, vector<std::string> *timezones,
         return;
     }
 
-    if (!e)
-    {
-        MegaTimeZoneDetails *tzDetails = new MegaTimeZoneDetailsPrivate(timezones, timezoneoffsets, defaulttz);
-        request->setTimeZoneDetails(tzDetails);
-    }
+    request->setTimeZoneDetails(tzDetails);
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: Done!

File: src/megaapi_impl.cpp [hunk 4]
```diff
@@ -29486,5 +29690,500 @@ int MegaTimeZoneDetailsPrivate::getDefault() const
     return defaultTimeZone;
 }
 
+MegaPushNotificationSettingsPrivate::MegaPushNotificationSettingsPrivate(const string &settingsJSON)
+{
+    JSON json;
+    json.begin(settingsJSON.c_str() + 1);
+    std::string name = json.getname();
+    while (name != "")
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: We'll try to keep the `getname()`, but checking for inconsistent value or parsing issues. If so, set a flag to indicate the JSON was invalid.

File: src/megaapi_impl.cpp [hunk 5]
```diff
@@ -29486,5 +29690,500 @@ int MegaTimeZoneDetailsPrivate::getDefault() const
...
+            }
+        }
+
+        name = json.getname();
+    }
+}
+
+MegaPushNotificationSettingsPrivate::MegaPushNotificationSettingsPrivate()
+{
+    mGlobalDND = -1;
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: done!

File: src/megaapi_impl.cpp [hunk 6]
```diff
@@ -2942,6 +2948,21 @@ void MegaRequestPrivate::setMegaFolderInfo(const MegaFolderInfo *folderInfo)
     this->folderInfo = folderInfo ? folderInfo->copy() : NULL;
 }
 
+MegaPushNotificationSettings *MegaRequestPrivate::getMegaPushNotificationSettings() const
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: Done!

File: src/megaapi_impl.cpp [hunk 7]
```diff
@@ -29486,5 +29690,500 @@ int MegaTimeZoneDetailsPrivate::getDefault() const
...
+                        else if (scheduleSubTitle == "start")
+                        {
+                            mGlobalScheduleStart = scheduleJson.getint();
+                        }
+                        else if (scheduleSubTitle == "end")
+                        {
+                            mGlobalScheduleEnd = scheduleJson.getint();
+                        }
+
+                        scheduleSubTitle = scheduleJson.getname();
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: done

File: src/megaapi_impl.cpp [hunk 8]
```diff
@@ -29486,5 +29690,500 @@ int MegaTimeZoneDetailsPrivate::getDefault() const
...
+            JSON jsonChats;
+            jsonChats.begin(globalObject.c_str() + 1);
+            std::string subname = jsonChats.getname();
+            assert(subname == "dnd");
+            mGlobalChatsDND = jsonChats.getint();
+        }
+        else
+        {
+            unsigned int base64Size = MegaClient::CHATHANDLE * 4 / 3 + 4;
+            assert(name.length() != base64Size);
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: done!

File: src/megaapi_impl.cpp [hunk 9]
```diff
@@ -29551,5 +29735,505 @@ int MegaTimeZoneDetailsPrivate::getDefault() const
...
+}
+
+MegaPushNotificationSettingsPrivate::~MegaPushNotificationSettingsPrivate()
+{
+
+}
+
+bool MegaPushNotificationSettingsPrivate::isGlobalEnabled() const
+{
+    return (mGlobalDND == -1 || (mGlobalDND > 0 && mGlobalDND < m_time(NULL)));
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp]: Should this check consider the `nsch` (notification schedule) too? IMHO, based exclusively on the DND might be confusing. Let's discuss about it tomorrow, @alber2510

File: src/megaapi_impl.cpp [hunk 10]
```diff
@@ -29551,5 +29735,505 @@ int MegaTimeZoneDetailsPrivate::getDefault() const
...
+
+const char *mega::MegaPushNotificationSettingsPrivate::getGlobalScheduleTimezone() const
+{
+    return MegaApi::strdup(mGlobalScheduleTimezone.c_str());
+}
+
+bool MegaPushNotificationSettingsPrivate::isChatEnabled(MegaHandle chatid) const
+{
+    std::map<uint64_t, time_t>::const_iterator it = mChatDND.find(chatid);
+    return (it == mChatDND.end() || it->second != 0);
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp]: Should this check consider the `an` too?

File: src/megaapi_impl.cpp [hunk 11]
```diff
@@ -29551,5 +29735,505 @@ int MegaTimeZoneDetailsPrivate::getDefault() const
...
+    if (!isGlobalEnabled())
+    {
+        LOG_warn << "disableGlobalDnd(): global notifications were disabled. Now are enabled";
+    }
+    mGlobalDND = -1;
+}
+
+void MegaPushNotificationSettingsPrivate::setGlobalSchedule(int start, int end, const char *timezone)
+{
+    if (start <= -1 || end <= -1 || !timezone || !timezone[0])
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp]: Isn't a wrong argument `start == end`??

File: src/megaapi_impl.cpp [hunk 12]
```diff
@@ -29551,5 +29735,505 @@ int MegaTimeZoneDetailsPrivate::getDefault() const
...
+        return;
+    }
+
+    if (enable)
+    {
+        mChatDND.erase(chatid);
+    }
+    else    // disable
+    {
+        mChatDND[chatid] = 0;
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp]: Should we check here if `always-notify` is set and, if so, disable it?

File: src/megaapi_impl.cpp (hunk 10 above)
Comment (alber2510) [reply src/megaapi_impl.cpp]: I don't think so, if chat is notifiable is managed by bool MegaApi::isChatNotifiable(MegaHandle chatid), with this function only managed if we have to enable/disable the UI
I've added a commit to modify the documentation

**Pull request merged**